pallets/phala/src/basepool.rs

```diff
@@ -0,0 +1,645 @@
+			let price = self.share_price()?;
+			let locked = nft.stakes;
+			let new_locked = bmul(nft.shares, &price);
+			// Double check the new_locked won't exceed the original locked
+			let new_locked = new_locked.min(locked);
+			// When only dust remaining in the pool, we include the dust in the slash amount
+			let (new_locked, _) = extract_dust(new_locked);
+			nft.stakes = new_locked;
+			// The actual slashed amount. Usually slash will only cause the share price decreasing.
+			// However in some edge case (i.e. the pool got slashed to 0 and then new contribution
+			// added), the locked amount may even become larger
+			Some(locked - new_locked)
+		}
+
+		// Distributes additional rewards to the current share holders.
+		//
+		// Additional rewards contribute to the face value of the pool shares. The value of each
+		// share effectively grows by (rewards / total_shares).
+		//
+		// Warning: `total_reward` mustn't be zero.
+		pub fn distribute_reward<T: Config>(
+			&mut self,
+			rewards: Balance,
+			need_update_free_stake: bool,
+		) where
+			T: pallet_uniques::Config<CollectionId = CollectionId, ItemId = NftId>,
+			BalanceOf<T>:
+				sp_runtime::traits::AtLeast32BitUnsigned + Copy + FixedPointConvert + Display,
+			T: mining::Config<Currency = <T as Config>::Currency>,
+			T: frame_system::Config<AccountId = AccountId>,
+			AccountId: PartialEq + Clone,
+		{
+			self.total_stake += rewards;
+			if need_update_free_stake {
+				self.free_stake += rewards;
+			}
+
+			for vault_staker in &self.vault_stakers {
+				let mut vault = ensure_vault::<T>(*vault_staker)
+					.expect("vault in vault_stakers should always exist: qed.");
+				let nft_id = Pallet::<T>::merge_or_init_nft_for_staker_inner(
+					self.cid,
+					vault.user_id.clone(),
+				)
+				.expect("merge nft shoule always success: qed.");
+				let nft = Pallet::<T>::get_nft_attr_inner(self.cid, nft_id)
+					.expect("get nft attr should always success: qed.");
+				let vault_shares = nft.shares.to_fixed();
+				let stake_ratio = match vault_shares.checked_div(self.total_shares.to_fixed()) {
+					Some(ratio) => BalanceOf::<T>::from_fixed(&ratio),
+					None => continue,
+				};
+				let settled_stake = bmul(stake_ratio, &rewards.to_fixed());
+				vault.basepool.distribute_reward::<T>(settled_stake, false);
+			}
+		}
+	}
+
+	pub fn pallet_id<T>() -> T
+	where
+		T: Encode + Decode,
+	{
+		(b"basepool")
+			.using_encoded(|b| T::decode(&mut TrailingZeroInput::new(b)))
+			.expect("Decoding zero-padded account id should always succeed; qed")
+	}
+
+	impl<T: Config> Pallet<T>
+	where
+		T: mining::Config<Currency = <T as Config>::Currency>,
+		BalanceOf<T>: sp_runtime::traits::AtLeast32BitUnsigned + Copy + FixedPointConvert + Display,
+		T: pallet_uniques::Config<CollectionId = CollectionId, ItemId = NftId>,
+	{
+		#[frame_support::transactional]
+		pub fn contribute<F>(
+			base_pool: &mut BasePool<T::AccountId, BalanceOf<T>>,
```

Comment:
In the context of this file, we can basically rename all the `base_pool` to just `pool`. I also suggest to rename `userid` to `account_id`.

Reply:
done